doc/man1/openssl-verification-options.pod

```diff
@@ -151,10 +151,13 @@ The first step is to check that each certificate is well-formed.
 Part of these checks are enabled only if the B<-x509_strict> option is given.
 
 The second step is to check the extensions of every untrusted certificate
-for consistency with the given purpose.
-If the B<-purpose> option is not included then no such checks are done.
-The target certificate must not have an EKU extension that is incompatible with
-the given purpose, and all other certificates must be valid CA certificates.
+for consistency with the supplied purpose.
+If the B<-purpose> option is not given then no such checks are done
+except for SSL/TLS connection setup,
+where by default C<sslserver> or C<sslclient>, are checked.
+The target or "leaf" certificate, as well as any other untrusted certificates,
```

Comment:
The EKU restrictions are also applied to the trusted certificates, not just the untrusted ones, unless someone changed that recently (which I would question as likely unwise).

Reply:
Interesting point - can you substantiate this from the sources or tests?
I am not aware of any changes to the trust checking during the last two years.
Conceptually, AFAIR, RFC 5280 does not specify any checks to trust anchors.

Reply:
> Interesting point - can you substantiate this from the sources or tests?
> I am not aware of any changes to the trust checking during the last two years.
> Conceptually, AFAIR, RFC 5280 does not specify any checks to trust anchors.

The trust store does not just contain trust anchors, it can also contain preferred intermediate certificates, which are not trust anchors, and are subject to various constraints, path length constraints, ...

But even, and especially trust anchors (RFC5280 notwithstanding) need to be subjected to appropriate use constraints, which can either be specified via "TRUSTED CERTIFICATE" external EKUs, or perhaps EKUs in the self-signed cert (though of course rarely seen in practice).

The check chain certs code applies the purpose checks to the entire chain, not just the untrusted certs.  Certainly in OpenSSL 1.1.1, if somebody changed this in 3.0, I think that'd be a mistake.

Reply:
Looks to me like check_trust() is still checked for certs in the trust store, and if some intermediate has explicit EKUs, they'd be enforced.

Reply:
I don't have time lately to keep track of all the changes you're making, so you need to be better at paying attention to detail.

The certificate verification code is somewhat irreducibly complex, and one needs to be extremely cautious and detail oriented when working on it. Read, re-read, take notes, look up and down the call tree, ... and if something is unclear, don't change it.

Reply:
Right now, I do not have the time to look deeper, but I have the impression that the issue boils down to what one means by "trusted certificates". If those refer to members of the trust store, I agree they are not yet really trusted, but are subject to further filtering by trust constraints etc., before they become actual trust anchors in the sense of 5280.

It looks like in the docs a statement is (still) missing that also members of the trust store are subject to purpose constraints.
Yet this would need to be said in a different paragraph than the one on which your comment was added to, because this paragraph pertains to untrusted certs (non-TAs) only.

Reply:
Users are already exposed to a notion of trust as meaning "in the trust store", e.g. via the "-trusted" and "-untrusted" flags to verify(1), and the exposition that talks about chain construction as looking at untrusted certificates from the peer when there's no match in the trust store.  Thus "trusted" is broader "trust-anchor".

I don't recall whether your proposed change to stop chain construction at the first matching trust-store certificate ended getting merged or not, and I remain sceptical that it would be (or was) a good idea to merge it.  In any case the text needs to be pretty clear about what is meant here, because it sure seemed like it was talking about CA certs in the trust store generally, rather that specifically trust anchors.

Care is required in defining clear terms and using them consistently.  Good luck!

Reply:
> The EKU restrictions are also applied to the trusted certificates, not just the untrusted ones, unless someone changed that recently (which I would question as likely unwise).

@vdukhovni, it turns out that your concern and the exchange we had this morning could have been avoided by simply reading on:
> The third step is to check the trust settings on the last certificate
(which typically is a self-signed root CA certificate).
It must be trusted for the given use.
For compatibility with previous versions of OpenSSL, a self-signed certificate
with no trust attributes is considered to be valid for all uses.

Reply:
Sorry, no.  The text appears to suggest that we don't check the EKUs of trusted (from the trust store), but not *externally* decorated with explicit trust EKUs (i.e. wrapped as a TRUSTED CERTIFICATE).  But in fact we do check such EKUs.

Reply:
Ah, looks like you are after the distinction between 
* EKUs used for OpenSSL trust attributes (decorations external to X.509) and 
* EKUs within X.509 cert extensions, which apparently are referred to as "purpose" in the OpenSSL sources, 

right?
This distinction is still not clear from the documentation.
The `Explicit trust/purpose combinations` tests in `25-test_verify.t` confirm that 
both forms of EKUs are checked for the last cert in a chain, which typically is a "trusted" root cert.